### DISCUSSION TITLE
M1 brew g++ "addend too big for relocation"

### DISCUSSION THREAD

**Discussion by raykong1** [Homebrew/discussions#2687] [Everyday usage] [answered]
When initializing an array of vectors in c++ of size 500000 globally (``std::vector<int> arr[500000]``), g++ throws ``error: addend too big for relocation``

example code:
```cpp
#include <vector>
std::vector<int> arr[500000];

int main(){
	return 0;
}
```

compiler flag:
```
g++ -Wl,-stack_size,20000000 -std=gnu++17 -O0 -DLOCAL
```

**Comment by SMillerDev** [comment, marked as answer]
GCC is unsupported on M1, we use an unofficial fork until it is supported to give people at least some GCC on Apple Silicon. Try checking the issues at https://github.com/fxcoudert/gcc

**Reply by jeffhammond** [reply]
Small problem: https://github.com/fxcoudert/gcc doesn't show the issues tab.

**Reply by Bo98** [reply]
https://github.com/iains/gcc-darwin-arm64 is the correct repo, but the issue has already been identified and fixed (https://github.com/iains/gcc-darwin-arm64/commit/040bb280be93df51c344c0720c241b658609ac61) but our GCC 11.2 build was released 2 days before that.

We're preparing for GCC 12 (releasing on 6th May), so the fix will be included in that build.

**Comment by Abner-Barros** [comment]
I have encountered the same issue while using the gfortran compiler on the M1. 

I am getting the two following errors: 
`/var/folders/73/6w_6mgms3td_0d0xyvvlmfqw0000gn/T//ccI1kPjw.s:96:2: error: addend too big for relocation`
and
`/var/folders/73/6w_6mgms3td_0d0xyvvlmfqw0000gn/T//ccI1kPjw.s:98:2: error: fixup value out of range`

**Comment by HiggsBoson3310** [comment]
I just ran into this problem while running a simple example of matrix diagonalization. It was solved by using allocatable arrays. 

This was the code that gave the error (I don't know if it is a minimal example)

```
program yalemult
    use matmod
    implicit none
    integer, parameter :: nnt=3000
    complex*16 :: M(nnt,nnt), V(nnt,nnt), vec(nnt)
    real*8 :: D(nnt)
    real*8 :: xx, yy, start, end,  mc, sparse
    integer :: i, j
    write(6,*) "Using a size of", nnt
    do i=1,nnt
        call random_number(xx)
        call random_number(yy)
        vec(i) = COMPLEX(xx,yy)
        do j=i,nnt
            call random_number(mc)
            if(mc.ge.0.92) then
                call random_number(xx)
                call random_number(yy)
                M(i,j) = COMPLEX(xx,yy)
                M(j,i) = CONJG(M(i,j))
            else
                if(i.eq.j) then 
                    sparse = sparse + 1
                else 
                    sparse = sparse + 2
                endif
            endif
        end do
    end do
    write(6,*) "The matrix is ", sparse/nnt**2 * 100, "% sparse"
    call cpu_time(start)
    call zdiag(nnt,M,D,V)
    call cpu_time(end)
    write(6,*) "Diagonalization took ", end-start, " seconds"
    call cpu_time(start)
    vec = matmul(CONJG(TRANSPOSE(V)),vec)

        do i=1,nnt
            vec(i) = EXP(-1.d0*COMPLEX(0,1.d0)*D(i))*vec(i)
        end do

        vec = matmul(V, vec)
    call cpu_time(end)
    write(6,*) "A single step took ", end-start, "seconds"
end program yalemult
```

sub routine ``zdiag`` is just a subroutine that calls the appropriate LAPACK functions to do the diagonalization. Using allocatable arrays for M, V, vec and D solved the problem.

**Reply by DoddyPhysics** [reply]
I encountered and solved this problem by reinstalling LAPACK with brew.
